File: java/src/test/java/algorithm/tests/geometry/SolidsTest.java
package algorithm.tests.geometry;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.function.Executable;

import algorithm.geometry.solids.Cylinder;
import algorithm.geometry.solids.Sphere;
import algorithm.tests.fixtures.Helper;
import algorithm.utils.errors.parameters.NegativeParameterError;
import algorithm.utils.errors.parameters.ZeroParameterError;

@DisplayName("Solids Figures Test Suite")
public class SolidsTest {
    @Nested
    @DisplayName("Cylinder Test Suite")
    class CylinderTest {
        @Test
        @DisplayName("Cylinder Negative or Zero Radius Test")
        void newCylinder_negativeOrZeroRadius_throwsException() {
            Executable negativeCylinder = () -> new Cylinder(-1, 2);
            Executable zeroCylinder = () -> new Cylinder(0, 2);
            assertThrows(NegativeParameterError.class, negativeCylinder);
            assertThrows(ZeroParameterError.class, zeroCylinder);
        }

        @Test
        @DisplayName("Cylinder Negative or Zero Height Test")
        void newCylinder_negativeOrZeroHeight_throwsException() {
            Executable negativeCylinder = () -> new Cylinder(2, -1);
            Executable zeroCylinder = () -> new Cylinder(2, 0);
            assertThrows(NegativeParameterError.class, negativeCylinder);
            assertThrows(ZeroParameterError.class, zeroCylinder);
        }

        @Test
        @DisplayName("Cylinder Radius Test")
        void cylinderRadius_returnsCorrectValue() throws Exception {
            Cylinder cylinder = new Cylinder(3.0, 5.0);
            double actual = cylinder.getRadius();
            double expected = 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Cylinder Height Test")
        void cylinderHeight_returnsCorrectValue() throws Exception {
            Cylinder cylinder = new Cylinder(3.0, 5.0);
            double actual = cylinder.getHeight();
            double expected = 5.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Cylinder Volume Test")
        void cylinderVolume_returnsCorrectValue() throws Exception {
            Cylinder cylinder = new Cylinder(3.0, 5.0);
            double actual = cylinder.volume();
            double expected = Math.PI * Math.pow(3.0, 2) * 5.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Cylinder Surface Area Test")
        void cylinderSurfaceArea_returnsCorrectValue() throws Exception {
            Cylinder cylinder = new Cylinder(3.0, 5.0);
            double actual = cylinder.surfaceArea();
            double expected = 2 * Math.PI * 3.0 * (3.0 + 5.0);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }
    }

    @Nested
    @DisplayName("Sphere Test Suite")
    class SphereTest {
        @Test
        @DisplayName("Sphere Negative or Zero Radius Test")
        void newSphere_negativeOrZeroRadius_throwsException() {
            Executable negativeSphere = () -> new Sphere(-1);
            Executable zeroSphere = () -> new Sphere(0);
            assertThrows(NegativeParameterError.class, negativeSphere);
            assertThrows(ZeroParameterError.class, zeroSphere);
        }

        @Test
        @DisplayName("Sphere Radius Test")
        void sphereRadius_returnsCorrectValue() throws Exception {
            Sphere sphere = new Sphere(3.0);
            double actual = sphere.getRadius();
            double expected = 3.0;
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Sphere Volume Test")
        void sphereVolume_returnsCorrectValue() throws Exception {
            Sphere sphere = new Sphere(3.0);
            double actual = sphere.volume();
            double expected = (4.0 / 3.0) * Math.PI * Math.pow(3.0, 3);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }

        @Test
        @DisplayName("Sphere Surface Area Test")
        void sphereSurfaceArea_returnsCorrectValue() throws Exception {
            Sphere sphere = new Sphere(3.0);
            double actual = sphere.surfaceArea();
            double expected = 4 * Math.PI * Math.pow(3.0, 2);
            assertEquals(expected, actual, Helper.errorMessage(actual, expected));
        }
    }
}

File: rust/tests/geometry/solids_test.rs
use std::f64::consts::PI;

use algorithm::geometry::Solid;
use algorithm::geometry::solids::{Cylinder, Sphere};
use algorithm::utils::errors::parameters::{NegativeParameterError, ZeroParameterError};

use crate::fixtures::helper;

#[cfg(test)]
mod cylinder_test {
    use super::*;

    #[test]
    fn cylinder_new_negative_or_zero_radius_returns_error() {
        let negative_cylinder = Cylinder::new(-1.0, 2.0);
        let zero_cylinder = Cylinder::new(0.0, 2.0);
        assert!(matches!(negative_cylinder, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_cylinder, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn cylinder_new_negative_or_zero_height_returns_error() {
        let negative_cylinder = Cylinder::new(2.0, -1.0);
        let zero_cylinder = Cylinder::new(2.0, 0.0);
        assert!(matches!(negative_cylinder, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_cylinder, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn cylinder_radius_returns_correct_value() {
        let cylinder = Cylinder::new(3.0, 4.0).unwrap();
        let actual = cylinder.radius();
        let expected = 3.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn cylinder_height_returns_correct_value() {
        let cylinder = Cylinder::new(3.0, 4.0).unwrap();
        let actual = cylinder.height();
        let expected = 4.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn cylinder_surface_area_returns_correct_value() {
        let cylinder = Cylinder::new(3.0, 4.0).unwrap();
        let actual = cylinder.surface_area();
        let expected = 2.0 * PI * 3.0 * (3.0 + 4.0);
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn cylinder_volume_returns_correct_value() {
        let cylinder = Cylinder::new(3.0, 4.0).unwrap();
        let actual = cylinder.volume();
        let expected = PI * 3.0f64.powi(2) * 4.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }
}

#[cfg(test)]
mod sphere_test {
    use super::*;

    #[test]
    fn sphere_new_negative_or_zero_radius_returns_error() {
        let negative_sphere = Sphere::new(-1.0);
        let zero_sphere = Sphere::new(0.0);
        assert!(matches!(negative_sphere, Err(e) if e.is::<NegativeParameterError>()));
        assert!(matches!(zero_sphere, Err(e) if e.is::<ZeroParameterError>()));
    }

    #[test]
    fn sphere_radius_returns_correct_value() {
        let sphere = Sphere::new(3.0).unwrap();
        let actual = sphere.radius();
        let expected = 3.0;
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn sphere_surface_area_returns_correct_value() {
        let sphere = Sphere::new(3.0).unwrap();
        let actual = sphere.surface_area();
        let expected = 4.0 * PI * 3.0f64.powi(2);
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }

    #[test]
    fn sphere_volume_returns_correct_value() {
        let sphere = Sphere::new(3.0).unwrap();
        let actual = sphere.volume();
        let expected = (4.0 / 3.0) * PI * 3.0f64.powi(3);
        assert_eq!(actual, expected, "{}", helper::error_message(actual, expected));
    }
}
